// @lumigo/node-core/lib/common/configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getChainedServicesMaxWidth = exports.getChainedServicesMaxDepth = exports.DEFAULT_CHAINED_SERVICES_MAX_WIDTH = exports.CHAINED_SERVICES_MAX_WIDTH = exports.DEFAULT_CHAINED_SERVICES_MAX_DEPTH = exports.CHAINED_SERVICES_MAX_DEPTH = void 0;
exports.CHAINED_SERVICES_MAX_DEPTH = 'LUMIGO_CHAINED_SERVICES_MAX_DEPTH';
exports.DEFAULT_CHAINED_SERVICES_MAX_DEPTH = 3;
exports.CHAINED_SERVICES_MAX_WIDTH = 'LUMIGO_CHAINED_SERVICES_MAX_WIDTH';
exports.DEFAULT_CHAINED_SERVICES_MAX_WIDTH = 5;
const getChainedServicesMaxDepth = () => {
    if (process.env[exports.CHAINED_SERVICES_MAX_DEPTH])
        return parseFloat(process.env[exports.CHAINED_SERVICES_MAX_DEPTH]);
    return exports.DEFAULT_CHAINED_SERVICES_MAX_DEPTH;
};
exports.getChainedServicesMaxDepth = getChainedServicesMaxDepth;
const getChainedServicesMaxWidth = () => {
    if (process.env[exports.CHAINED_SERVICES_MAX_WIDTH])
        return parseFloat(process.env[exports.CHAINED_SERVICES_MAX_WIDTH]);
    return exports.DEFAULT_CHAINED_SERVICES_MAX_WIDTH;
};
exports.getChainedServicesMaxWidth = getChainedServicesMaxWidth;
