// @lumigo/node-core/lib/common/triggers/trigger-parsers/appsync-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppSyncMessageParser = void 0;
const trigger_parser_base_1 = require("./trigger-parser-base");
const generalUtils_1 = require("../../generalUtils");
const message_trigger_enum_1 = require("../message-trigger.enum");
class AppSyncMessageParser extends trigger_parser_base_1.MessageTriggerParser {
    constructor() {
        super(...arguments);
        this._shouldHandle = (message) => {
            var _a, _b, _c;
            return (_c = (_b = (_a = message === null || message === void 0 ? void 0 : message.request) === null || _a === void 0 ? void 0 : _a.headers) === null || _b === void 0 ? void 0 : _b.host) === null || _c === void 0 ? void 0 : _c.includes('appsync-api');
        };
        this.handle = (message, targetId) => {
            const { host, 'x-amzn-trace-id': traceId } = message.request.headers;
            return {
                id: (0, generalUtils_1.getRandomId)(),
                targetId: targetId,
                triggeredBy: message_trigger_enum_1.MessageTrigger.AppSync,
                fromMessageIds: [traceId.split('=')[1]],
                extra: {
                    api: host,
                },
            };
        };
    }
}
exports.AppSyncMessageParser = AppSyncMessageParser;
