// @lumigo/node-core/lib/common/triggers/trigger-parsers/kinesis-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KinesisMessageParser = void 0;
const trigger_parser_base_1 = require("./trigger-parser-base");
const generalUtils_1 = require("../../generalUtils");
const message_trigger_enum_1 = require("../message-trigger.enum");
class KinesisMessageParser extends trigger_parser_base_1.MessageTriggerParser {
    constructor() {
        super(...arguments);
        this._shouldHandle = (message) => {
            var _a, _b;
            return ((_b = (_a = message === null || message === void 0 ? void 0 : message.Records) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.eventSource) === 'aws:kinesis';
        };
        this.handle = (message, targetId) => {
            const arn = message.Records[0].eventSourceARN;
            const messageIds = (message.Records || [])
                .map((record) => record.kinesis.sequenceNumber)
                .filter((recordSequenceNumber) => recordSequenceNumber != null);
            return {
                id: (0, generalUtils_1.getRandomId)(),
                targetId: targetId,
                triggeredBy: message_trigger_enum_1.MessageTrigger.Kinesis,
                fromMessageIds: messageIds,
                extra: { arn },
            };
        };
    }
}
exports.KinesisMessageParser = KinesisMessageParser;
