// @lumigo/node-core/lib/common/triggers/trigger-parsers/s3-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3MessageParser = void 0;
const trigger_parser_base_1 = require("./trigger-parser-base");
const generalUtils_1 = require("../../generalUtils");
const message_trigger_enum_1 = require("../message-trigger.enum");
class S3MessageParser extends trigger_parser_base_1.MessageTriggerParser {
    constructor() {
        super(...arguments);
        this._shouldHandle = (message) => {
            var _a, _b;
            return ((_b = (_a = message === null || message === void 0 ? void 0 : message.Records) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.eventSource) === 'aws:s3';
        };
        this.handle = (message, targetId) => {
            var _a, _b, _c, _d, _e, _f, _g;
            return {
                id: (0, generalUtils_1.getRandomId)(),
                targetId: targetId,
                triggeredBy: message_trigger_enum_1.MessageTrigger.S3,
                fromMessageIds: [(_c = (_b = (_a = message === null || message === void 0 ? void 0 : message.Records) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.responseElements) === null || _c === void 0 ? void 0 : _c['x-amz-request-id']],
                extra: {
                    arn: (_g = (_f = (_e = (_d = message === null || message === void 0 ? void 0 : message.Records) === null || _d === void 0 ? void 0 : _d[0]) === null || _e === void 0 ? void 0 : _e.s3) === null || _f === void 0 ? void 0 : _f.bucket) === null || _g === void 0 ? void 0 : _g.arn,
                },
            };
        };
    }
}
exports.S3MessageParser = S3MessageParser;
