// @lumigo/node-core/lib/common/triggers/trigger-parsers/sns-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnsMessageParser = void 0;
const trigger_parser_base_1 = require("./trigger-parser-base");
const message_trigger_enum_1 = require("../message-trigger.enum");
const generalUtils_1 = require("../../generalUtils");
class SnsMessageParser extends trigger_parser_base_1.MessageTriggerParser {
    constructor() {
        super(...arguments);
        this.INNER_IDENTIFIER = 'SimpleNotificationService';
        this._shouldHandle = (message) => {
            var _a, _b;
            return (((_b = (_a = message === null || message === void 0 ? void 0 : message.Records) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.EventSource) === 'aws:sns' ||
                ((message === null || message === void 0 ? void 0 : message.Type) === 'Notification' && (message === null || message === void 0 ? void 0 : message.TopicArn)));
        };
        this.handle = (message, targetId) => {
            var _a, _b;
            const record = ((_b = (_a = message === null || message === void 0 ? void 0 : message.Records) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.Sns) || message;
            return {
                id: (0, generalUtils_1.getRandomId)(),
                targetId: targetId,
                triggeredBy: message_trigger_enum_1.MessageTrigger.SNS,
                fromMessageIds: [record.MessageId],
                extra: {
                    arn: record.TopicArn,
                },
            };
        };
    }
}
exports.SnsMessageParser = SnsMessageParser;
