// @lumigo/node-core/lib/common/triggers/trigger-parsers/sqs-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqsMessageParser = void 0;
const trigger_parser_base_1 = require("./trigger-parser-base");
const message_trigger_enum_1 = require("../message-trigger.enum");
const generalUtils_1 = require("../../generalUtils");
class SqsMessageParser extends trigger_parser_base_1.MessageTriggerParser {
    constructor() {
        super(...arguments);
        this._shouldHandle = (message) => {
            var _a, _b;
            return ((_b = (_a = message === null || message === void 0 ? void 0 : message.Records) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.eventSource) === 'aws:sqs';
        };
        this.handle = (message, targetId) => {
            const arn = message.Records[0].eventSourceARN;
            const messageIds = [];
            message.Records.forEach((record) => record.messageId && messageIds.push(record.messageId));
            return {
                id: (0, generalUtils_1.getRandomId)(),
                targetId: targetId,
                triggeredBy: message_trigger_enum_1.MessageTrigger.SQS,
                fromMessageIds: messageIds,
                extra: { arn },
            };
        };
        this.extractInner = (message) => {
            return message.Records.map((record) => record.body).filter((body) => !!body);
        };
    }
}
exports.SqsMessageParser = SqsMessageParser;
